codalab/bin/ws_server.py

```diff
@@ -1,75 +1,148 @@
+    def __init__(self, timeout_seconds: float = 60):
+        self._lock = threading.Lock()
+        self._time_since_locked: float
+        self._timeout: float = timeout_seconds
 
+    def acquire(self, blocking=True, timeout=-1):
+        acquired = self._lock.acquire(blocking, timeout)
+        if acquired:
+            self._time_since_locked = time.time()
+        return acquired
+
+    def locked(self):
+        return self._lock.locked()
+
+    def release(self):
+        self._lock.release()
+
+    def timeout(self):
+        return time.time() - self._time_since_locked > self._timeout
+
+    def release_if_timeout(self):
+        if self.locked() and self.timeout():
+            self.release()
 
-async def worker_handler(websocket, worker_id):
-    """Handles routes of the form: /worker/{id}. This route is called when
-    a worker first connects to the ws-server, creating a connection that can
-    be used to ask the worker to check-in later.
-    """
-    # runs on worker connect
-    worker_to_ws[worker_id] = websocket
-    logger.warning(f"Connected to worker {worker_id}!")
 
+worker_to_ws: Dict[str, Dict[str, Any]] = defaultdict(dict)  # Maps worker to socket ID to websocket
+ws_to_lock: Dict[str, Dict[str, TimedLock]] = defaultdict(dict)
+ACK = b'a'
+logger = logging.getLogger(__name__)
+manager = CodaLabManager()
+bundle_model = manager.model()
+worker_model = manager.worker_model()
+server_secret = os.getenv("CODALAB_SERVER_SECRET")
+
+
+async def send_handler(server_websocket, worker_id):
+    """Handles routes of the form: /send/{worker_id}. This route is called by
+    the rest-server or bundle-manager when either wants to send a message/stream to the worker.
+    """
+    # Check if any websockets available
+    if worker_id not in worker_to_ws or len(worker_to_ws[worker_id]) == 0:
+        logger.warning(f"No websockets currently available for worker {worker_id}")
+        await server_websocket.close(
+            1013, f"No websockets currently available for worker {worker_id}"
+        )
+        return
+
+    # Authenticate server.
+    receieved_secret = await server_websocket.recv()
+    if receieved_secret != server_secret:
+        logger.warning("Server unable to authenticate.")
+        await server_websocket.close(1008, "Server unable to authenticate.")
+        return
+
+    # Send message from server to worker.
+    for socket_id, worker_websocket in random.sample(
+        worker_to_ws[worker_id].items(), len(worker_to_ws[worker_id])
+    ):
+        if ws_to_lock[worker_id][socket_id].acquire(blocking=False):
+            data = await server_websocket.recv()
+            await worker_websocket.send(data)
+            await server_websocket.send(ACK)
+            ws_to_lock[worker_id][socket_id].release()
+            return
+
+    logger.warning(f"All websockets for worker {worker_id} are currently busy.")
+    await server_websocket.close(1013, f"All websockets for worker {worker_id} are currently busy.")
```

Comment:
what is 1013?

Reply:
Hmm... weird. I see this listed in several other places (e.g. [here](https://github.com/Luka967/websocket-close-codes)) as `Try again later`, but not in [the RFC](https://www.rfc-editor.org/rfc/rfc6455#section-7.4).

Perhaps I should change this to a different status codes, but none of the ones listed in the RFC are fitting... what do you think?